### PR TITLE
Unify approach to expirable entities for the new store

### LINKED ISSUE #11947: Introduce expiration field for all entities that can expire and use it in all queries that load these entities
### Description

See: https://github.com/keycloak/keycloak/discussions/11166 for more details

### Discussion

https://github.com/keycloak/keycloak/discussions/11166

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by mhajas**
This PR is unifying the way how we approach expiration for the new store.

There are two main changes:
- All timestamps should be using milliseconds 
- All queries should include an EXPIRATION field to be sure we are not loading any expired entities into Kyecloak

Fixes: #11947

### CODE REVIEW

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/X509DirectGrantTest.java [hunk 1]
```diff
@@ -264,6 +267,8 @@ public void loginCertificateNotExpired() throws Exception {
 
     @Test
     public void loginCertificateExpired() throws Exception {
+        Assume.assumeFalse("Time offset is causing integer overflow. With the old store it works, because root authentication session has also timestamp overflown, this is not true for the new store so the test is failing.", keycloakUsingProviderWithId(AuthenticationSessionProvider.class, "map"));
+
         X509AuthenticatorConfigModel config =
                 new X509AuthenticatorConfigModel()
                     .setCertValidationEnabled(true)
```
Comment (ahus1) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/X509DirectGrantTest.java:274]: Would this test be re-enabled once this PR is ready, or would it be a separate issue?

File: model/map/src/main/java/org/keycloak/models/map/common/ExpirableEntity.java [hunk 2]
```diff
@@ -44,4 +44,18 @@ public interface ExpirableEntity extends AbstractEntity {
      * @param expiration a timestamp in milliseconds since The Epoch or {@code null} if this entity never expires.
      */
     void setExpiration(Long expiration);
+
+    /**
+     * Returns a point in time (timestamp in milliseconds since The Epoch) when this entity started its existence
+     *
+     * @return a timestamp in milliseconds since The Epoch or {@code null} when the time is unknown
+     */
+    Long getTimestamp();
```
Comment (ahus1) [inline model/map/src/main/java/org/keycloak/models/map/common/ExpirableEntity.java:53]: Looking at the docs of RootAuthenticationSessionModel#getTimestamp(), I read "Returns timestamp when the root authentication session was created or updated". 

Without so much knowledge about the code at hand, I wonder if this JavaDoc should be updated, and how the current code behaves with regards to updating this field. Looking at the code it seems that it expiry is always updated when the timestamp is updated as well.

Will an updated expiry time also be propagated to HotRod? Looking at the code I didn't find where the initial expiry or an update expiry is passed to HotRod. 

(applies also to setTimestamp())

**Review by ahus1 — CHANGES_REQUESTED** [review]
Thank you for these changes - as I have experienced before, timestamps touch a lot of places, and long vs. int makes it cumbersome in several places.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/X509DirectGrantTest.java (hunk 1 above)
Comment (mhajas) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/X509DirectGrantTest.java:274]: This is one of the things why I sent this just as a draft. Maybe I am missing something but it seems to enable this test we would need to make `Time.currentTime()` return long instead of int, which would require a lot of changes. That is why I decided to ignore this test for now and based on the discussion we have here, we will decide how to approach this.

To give you more details on why this is passing before my PR:
1. The test is moving in time 50 years to the future (after the year 2038 where integer timestamp overflows)
2. [This](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/services/managers/ClientSessionCode.java#L176) is the line where all of the magic happens :laughing:. `Time.currentTime()`  returns negative number. However, also timestamp is a negative number because we use `int` [here](https://github.com/keycloak/keycloak/blob/main/model/map/src/main/java/org/keycloak/models/map/authSession/MapRootAuthenticationSessionProvider.java#L102) (Not sure it is exactly the place, anyway it is one of the places where I changed `int` -> `long` in this PR).
3. Given this ^ the condition `lifespan > Time.currentTime() - timestamp` passes and the test works as expected. After my changes in this PR `timestamp` started to be positive number which breaks the condition.

That is why I think the only solution is to create some follow-up issue that would change `Time.currentTime` to long to make this test pass with the new store.

File: model/map/src/main/java/org/keycloak/models/map/common/ExpirableEntity.java (hunk 2 above)
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/common/ExpirableEntity.java:53]: Not sure what you mean by the first part of the comment, could you rephrase, please?

For the second part, HotRod currently doesn't use expiration in any way to set cache entry lifespan, this PR is just preparing storage to make this simply possible. This will be done in this follow-up task: https://github.com/keycloak/keycloak/issues/12068

Comment (ahus1) [reply model/map/src/main/java/org/keycloak/models/map/common/ExpirableEntity.java:53]: Reading that HotRod doesn't use the expiration as of now, I rephrase it as: Please update the JavaDoc from 

"Returns a point in time (timestamp in milliseconds since The Epoch) when this entity started its existence" 

to 

"Returns a point in time (timestamp in milliseconds since The Epoch) when the entity was created or updated."

as the field can be updated to indicate for example a refreshed session, like this is stated in the JavaDoc of RootAuthenticationSessionModel#getTimestamp.

Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/common/ExpirableEntity.java:53]: Hmm, I see it now thanks. That makes sense, I will do this.

**Review by martin-kanis — COMMENTED** [review]
With the existing approach (checking expiration in `entityToModelFunc()`) the expired entities are removed from the store. With the expiration field in `MCB` the expired entries will be filtered out and left in the store unless a storage is able to remove those (not implemented yet) or we have some periodic cleanup. How do we prevent the storage to grow until we have one of the aforementioned capabilities?

**Comment by mhajas** [comment]
This problem is there anyway. Entities that are never loaded to Keycloak stay in the storage even now. The solution is to do the garbage collection.

**Comment by mhajas** [comment]
Isn't the price we pay with each request for this solution too high if we can replace it with one SQL command running once a day at midnight? 

Also, if storage is not able to implement such garbage collection it can just skip the EXPIRATION field in each request and it will be cleaned up as you suggested.

**Review by martin-kanis — APPROVED** [review]
LGTM, thanks @mhajas.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/X509DirectGrantTest.java (hunk 1 above)
Comment (ahus1) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/X509DirectGrantTest.java:273]: @mhajas - please open a new issue for this. I'd assume one can move the time to 2037 (the year before 2038) to show the expiry of a certificate without relying on an overflow.

**Review by ahus1 — APPROVED** [review]
I approve the changes after a review. Please open a new issue for the test that is now disabled for map storage (see other comment)

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/X509DirectGrantTest.java (hunk 1 above)
Comment (ahus1) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/X509DirectGrantTest.java:273]: I once gave it a start for #11053 and it turned out to be so big that I doubt that someone will be able review it in the near future (unfortunately). Therefore recommend to look at that new issue independently. With what you wrote about which certificate would be involved here, that analysis should be part of that other issue.

**Review by hmlnarik — APPROVED** [review]
Thanks for the fix!
